doc/rst/developer/bestPractices/TestSystem.rst

```diff
@@ -103,86 +103,290 @@ example, given a directory containing two files:
+should be skipped when the environment variable on the left does not contain the
+contents on the right, while ``<=`` indicates the opposite - this is useful for
+imprecise matches, e.g. ``CHPL_HOST_PLATFORM >= cygwin`` would cause the test to
+run on both ``cygwin64`` and ``cygwin32``.
+
+The second form a ``.skipif`` or ``SKIPIF`` file can take is that of a script.
+This form is intended for conditions that require some computation to determine,
+or when the combination of conditions is necessary (i.e. this setting **and**
+this setting are required for the behavior we want to avoid).  The script can be
+in any commonly supported scripting language, usually bash or python.  The
+``.skipif`` or ``SKIPIF`` file must have executable permissions for this form to
+work.  Printing ``True`` to standard output will result in the test being
+skipped, while printing ``False`` will result in the test being run.
+
+For instance:
+
+``foo.skipif``
+
+  .. code-block::
+
+     #!/usr/bin/env python
+
+     import os
+     print(os.getenv('CHPL_TEST_PERF') == 'on' and
+           os.getenv('CHPL_ATOMICS') == 'locks')
+
+would cause the test to be skipped when performance testing is done with
+CHPL_ATOMICS=locks, but not ordinary performance testing, or correctness
+testing with CHPL_ATOMICS=locks
+
+Testing Different Behavior in Different Settings
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If a test is intended to work in all settings but will have slightly different
+behavior in some situations, it is appropriate to add additional ``.good`` files
+for those settings.  Some of these additional ``.good`` files will be used
+automatically by the testing system, while others will need to be specified
+explicitly in the ``.compopts`` or ``.execopts`` file for the test.
+
+``start_test`` automatically recognizes ``.good`` files with prefixes for
+``--no-local``, communication layer, locale model, and ``chpldoc``.  For example:
+
+- ``.comm-none.good``: used with CHPL_COMM=none (the unqualified ``.good`` file
+  will then apply for CHPL_COMM != none)
+- ``.no-local.good``: used with ``--no-local`` testing
+- ``.lm-numa.good``: used with CHPL_LOCALE_MODEL=numa
+- ``.doc.good``: used when testing ``chpldoc`` instead of ``chpl``
+
+Requests can be made for supporting additional formats if a common format
+does not appear to be covered automatically.
+
+If only some compilations or executions of a test need a specialized ``.good``
+file, a comment on the same line as the relevant options can be used.  For
+instance:
+
+``foo.execopts``
+
+  .. code-block::
+
+     --x=true # foo.true.good
+     --x=false # foo.false.good
+
+will compare test output to ``foo.true.good`` for the first execution and
+``foo.false.good`` for the second.
+
+Any line that is unlabeled will use the default ``.good`` for that test.
+Undefined behavior will occur when both the ``.compopts`` and ``.execopts``
+files specify a ``.good`` file in this way.
+
+A Performance Test
```

Comment:
I think this section would benefit from a "tell 'em what you're gonna tell 'em" introduction, covering (or at least mentioning) the steps from performance test to graph: compilation/execution, extraction of key values to ``.dat`` files, and graph generation.

Also a note along the lines of "Though your new checkout doesn't have any ``.dat`` files, the graphs on sourceforge/chapel-lang are generated from existing files.  So if you need to modify any ``.perfkeys`` or ``.graph`` files, please keep in mind the following so as to maintain compatibility: ..."